File: src/tinymonitor/alerts/smtp.py
import smtplib
from email.mime.text import MIMEText
from email.mime.multipart import MIMEMultipart
from datetime import datetime
from tinymonitor.alerts.provider import AlertProvider, logger
from tinymonitor.utils import system_info

class SMTPAlert(AlertProvider):
    def send(self, component, level, value, title, message):
        smtp_host = self.config.get("host")
        smtp_port = self.config.get("port", 587)
        smtp_user = self.config.get("user")
        smtp_pass = self.config.get("password")
        from_addr = self.config.get("from_addr")
        to_addrs = self.config.get("to_addrs") # Can be a list or a comma-separated string
        use_tls = self.config.get("use_tls", True)

        if not all([smtp_host, smtp_user, smtp_pass, from_addr, to_addrs]):
            logger.error(f"[{self.name}] Missing SMTP configuration (host, user, password, from_addr, or to_addrs)")
            return

        if isinstance(to_addrs, str):
            to_addrs = [addr.strip() for addr in to_addrs.split(",")]

        # System Info
        hostname = system_info.get_hostname()
        execution_time = datetime.now().strftime("%Y-%m-%d %H:%M:%S")
        ip_private = system_info.get_private_ip()
        ip_public = system_info.get_public_ip()
        load_avg = system_info.get_load_avg()
        uptime_pretty = system_info.get_uptime()

        # Email Subject
        subject = f"[{level}] {component} on {hostname} - {value}"

        # Email Body (HTML)
        html_content = f"""
        <html>
        <body>
            <h2>{title}</h2>
            <p><strong>Component:</strong> {component}</p>
            <p><strong>Value:</strong> {value}</p>
            <p><strong>Level:</strong> {level}</p>
            <hr>
            <h3>Machine Context</h3>
            <ul>
                <li><strong>Server:</strong> {hostname}</li>
                <li><strong>Private IP:</strong> {ip_private}</li>
                <li><strong>Public IP:</strong> {ip_public}</li>
                <li><strong>Load Avg:</strong> {load_avg}</li>
                <li><strong>Uptime:</strong> {uptime_pretty}</li>
                <li><strong>Time:</strong> {execution_time}</li>
            </ul>
        </body>
        </html>
        """

        msg = MIMEMultipart()
        msg['From'] = from_addr
        msg['To'] = ", ".join(to_addrs)
        msg['Subject'] = subject

        msg.attach(MIMEText(html_content, 'html'))

        try:
            server = smtplib.SMTP(smtp_host, smtp_port)
            server.ehlo()
            if use_tls:
                server.starttls()
                server.ehlo()
            
            server.login(smtp_user, smtp_pass)
            server.sendmail(from_addr, to_addrs, msg.as_string())
            server.close()
            logger.info(f"[{self.name}] Email sent to {to_addrs}")
        except Exception as e:
            logger.error(f"[{self.name}] Failed to send email: {e}")

File: internal/alerts/smtp.go
package alerts

import (
	"crypto/tls"
	"fmt"
	"net/smtp"
	"strings"
	"time"

	"github.com/Gu1llaum-3/tinymonitor/internal/config"
	"github.com/Gu1llaum-3/tinymonitor/internal/models"
	"github.com/Gu1llaum-3/tinymonitor/internal/utils"
)

// SMTPProvider sends alerts via email
type SMTPProvider struct {
	BaseProvider
	host     string
	port     int
	user     string
	password string
	fromAddr string
	toAddrs  []string
	useTLS   bool
}

// NewSMTPProvider creates a new SMTP provider
func NewSMTPProvider(cfg config.SMTPConfig) *SMTPProvider {
	return &SMTPProvider{
		BaseProvider: BaseProvider{
			ProviderName: "smtp",
			Enabled:      cfg.Enabled,
			Levels:       cfg.Levels,
			Rules:        cfg.Rules,
		},
		host:     cfg.Host,
		port:     cfg.Port,
		user:     cfg.User,
		password: cfg.Password,
		fromAddr: cfg.FromAddr,
		toAddrs:  cfg.ToAddrs,
		useTLS:   cfg.UseTLS,
	}
}

// Send sends an alert via email
func (p *SMTPProvider) Send(alert models.Alert) error {
	if p.host == "" || p.user == "" || p.password == "" || p.fromAddr == "" || len(p.toAddrs) == 0 {
		return fmt.Errorf("missing SMTP configuration")
	}

	// System Info
	hostname := utils.GetHostname()
	executionTime := time.Now().Format("2006-01-02 15:04:05")
	ipPrivate := utils.GetPrivateIP()
	ipPublic := utils.GetPublicIP()
	loadAvg := utils.GetLoadAvg()
	uptimePretty := utils.GetUptime()

	subject := fmt.Sprintf("[%s] %s on %s - %s", alert.Level, alert.Component, hostname, alert.Value)

	htmlContent := fmt.Sprintf(`
	<html>
	<body>
		<h2>%s</h2>
		<p><strong>Component:</strong> %s</p>
		<p><strong>Value:</strong> %s</p>
		<p><strong>Level:</strong> %s</p>
		<hr>
		<h3>Machine Context</h3>
		<ul>
			<li><strong>Server:</strong> %s</li>
			<li><strong>Private IP:</strong> %s</li>
			<li><strong>Public IP:</strong> %s</li>
			<li><strong>Load Avg:</strong> %s</li>
			<li><strong>Uptime:</strong> %s</li>
			<li><strong>Time:</strong> %s</li>
		</ul>
	</body>
	</html>`,
		alert.Title, alert.Component, alert.Value, alert.Level,
		hostname, ipPrivate, ipPublic, loadAvg, uptimePretty, executionTime)

	msg := fmt.Sprintf("From: %s\r\n"+
		"To: %s\r\n"+
		"Subject: %s\r\n"+
		"MIME-Version: 1.0\r\n"+
		"Content-Type: text/html; charset=UTF-8\r\n"+
		"\r\n%s",
		p.fromAddr, strings.Join(p.toAddrs, ", "), subject, htmlContent)

	addr := fmt.Sprintf("%s:%d", p.host, p.port)
	auth := smtp.PlainAuth("", p.user, p.password, p.host)

	if p.useTLS {
		return p.sendWithTLS(addr, auth, msg)
	}

	return smtp.SendMail(addr, auth, p.fromAddr, p.toAddrs, []byte(msg))
}

func (p *SMTPProvider) sendWithTLS(addr string, auth smtp.Auth, msg string) error {
	conn, err := tls.Dial("tcp", addr, &tls.Config{ServerName: p.host})
	if err != nil {
		// Try STARTTLS instead
		return p.sendWithSTARTTLS(addr, auth, msg)
	}
	defer conn.Close()

	client, err := smtp.NewClient(conn, p.host)
	if err != nil {
		return err
	}
	defer client.Close()

	if err := client.Auth(auth); err != nil {
		return err
	}

	if err := client.Mail(p.fromAddr); err != nil {
		return err
	}

	for _, toAddr := range p.toAddrs {
		if err := client.Rcpt(toAddr); err != nil {
			return err
		}
	}

	w, err := client.Data()
	if err != nil {
		return err
	}

	_, err = w.Write([]byte(msg))
	if err != nil {
		return err
	}

	err = w.Close()
	if err != nil {
		return err
	}

	LogInfo(p.ProviderName, "Email sent", "to", p.toAddrs)
	return client.Quit()
}

func (p *SMTPProvider) sendWithSTARTTLS(addr string, auth smtp.Auth, msg string) error {
	client, err := smtp.Dial(addr)
	if err != nil {
		return err
	}
	defer client.Close()

	if err := client.Hello("localhost"); err != nil {
		return err
	}

	if ok, _ := client.Extension("STARTTLS"); ok {
		config := &tls.Config{ServerName: p.host}
		if err := client.StartTLS(config); err != nil {
			return err
		}
	}

	if err := client.Auth(auth); err != nil {
		return err
	}

	if err := client.Mail(p.fromAddr); err != nil {
		return err
	}

	for _, toAddr := range p.toAddrs {
		if err := client.Rcpt(toAddr); err != nil {
			return err
		}
	}

	w, err := client.Data()
	if err != nil {
		return err
	}

	_, err = w.Write([]byte(msg))
	if err != nil {
		return err
	}

	err = w.Close()
	if err != nil {
		return err
	}

	LogInfo(p.ProviderName, "Email sent", "to", p.toAddrs)
	return client.Quit()
}
